db/seeds.rb

```diff
@@ -243,77 +239,146 @@ def generate_blackout
+  # save on the start date so validations run properly
+  travel_to(res.start_date) { res.save }
 end
 
-def mark_checked_out(res)
-  if rand < MISSED_CHANCE
-    res.checked_out = nil
-    res.status = 'missed'
-  else
-    res.status = 'checked_out'
-    res.checked_out = res.start_date
-    res.equipment_item = res.equipment_model.equipment_items.all.sample
-    res.checkout_handler_id = User.where('role = ? OR role = ? OR role = ?',
-                                         'checkout', 'admin', 'superuser'
-                                        ).all.sample.id
+def make_checked_out(res, _)
+  res.status = 'checked_out'
+  res.checked_out = res.start_date
+  res.equipment_item = res.equipment_model.equipment_items.all.sample
+  res.checkout_handler_id = User.where('role = ? OR role = ? OR role = ?',
+                                       'checkout', 'admin',
+                                       'superuser').all.sample.id
+end
+
+def make_returned(res, random = false)
+  # put reservation in the past
+  make_past(res, random)
+  # check out reservation
+  make_checked_out res
+
+  # check in reservation
+  res.status = 'returned'
+  len = res.due_date - res.start_date
+  offset = random ? rand(0...len).days : 0.days
+  res.checked_in = res.due_date - offset
+  res.checkin_handler_id = User.where('role = ? OR role = ? OR role = ?',
+                                      'checkout', 'admin',
+                                      'superuser').all.sample.id
+end
+
+def make_overdue(res, random = false)
+  # put reservation in the past
+  make_past(res, random)
+  # check out reservation
+  make_checked_out res
+
+  # ensure the reservation is actually overdue
+  # unlikely but necessary check due to randomness
+  if res.due_date >= Time.zone.today
+    len = res.due_date - res.start_date
+    res.due_date = Time.zone.today
+    res.due_date -= random ? rand(1.day..PAST_RANGE) : 1.day
+    res.start_date = res.due_date - len
+    travel_to(res.start_date) { res.save }
   end
+  res.overdue = true
+end
+
+def make_returned_overdue(res, random = false)
+  make_overdue(res, random)
+  len = Time.zone.today - res.due_date
+  offset = random ? rand(0...len).days : 0.days
+  res.status = 'returned'
+  res.checked_in = Time.zone.today - offset
+  res.checkin_handler_id = User.where('role = ? OR role = ? OR role = ?',
+                                      'checkout', 'admin',
+                                      'superuser').all.sample.id
+end
+
+def make_missed(res, random = false)
+  make_past(res, random)
+  res.status = 'missed'
+end
+
+def make_archived(res, random = false)
+  make_past(res, random) if random && rand < 0.5
+  res.status = 'archived'
+end
+
+def make_requested(res, random = false)
+  make_future(res, random) if random && rand < 0.5
+  res.status = 'requested'
+  res.flag(:request)
+end
+
+def make_denied(res, random = false)
+  make_requested(res, random)
+  res.status = 'denied'
+end
+
+def make_future(res, random = false)
```

Comment:
Could we put this with `make_past` just to be clearer (time-based methods vs status-based methods)? Overall the huge number of methods in this script make it a bit hard to keep track of what's going on. I don't know if there's a way to re-organize / refactor to make things cleaner, but it would definitely be nice.